qiskit/synthesis/permutation/permutation_lnn.py

```diff
@@ -0,0 +1,68 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Depth-efficient synthesis algorithm for Permutation gates."""
+
+from qiskit import QuantumCircuit
+from .permutation_utils import _inverse_pattern
+
+
+def synth_permutation_depth_lnn_kms(pattern):
+    """
+    Permutation synthesis algorithm from [1], Chapter 6.
+
+    Synthesizes any permutation of n qubits over linear nearest-neighbor
+    architecture using SWAP gates with depth at most n and size at most
+    n(n-1)/2 (where both depth and size are measured with respect to SWAPs).
+
+    Args:
+        pattern (Union[list[int], np.ndarray]): permutation pattern.
```

Comment:
I think we should probably document a bit about the meaning of the permutation pattern here in the documentation. You've got the comment below in the function body that explains why we need to invert the pattern but as a user reading the docs for this function I wouldn't know how to specify the pattern necessarily without looking at the in code comment.